Add documentation for monitoring API

dedemorton opened the pull request:
@suyograo and other reviewers: The monitoring API documentation for Alpha 1 (issue #4730 ) is ready to review. When looking through this PR, please look for and respond to inline comments that contain questions that begin with //QUESTION.

I've uploaded the built files so you can see how they look formatted:   https://logstashreference.firebaseapp.com/monitoring.html 

At some point, we might want to have a top-level container called "Monitoring Logstash" but that doesn't make sense until we have more content about monitoring (not just API docs).

ph commented on docs/static/monitoring-apis.asciidoc:
```diff
@@ -0,0 +1,312 @@
+
+=== Testing Logstash APIs in Sense
+
+//QUESTION: What do you think of this approach? I felt that the links for invoking Sense would be useful, but I wanted to avoid confusing users. If they don't notice this topic, they might still be confused. Do you think I should keep the links that launch Sense, or should I remove them?
```
I think sense is a great idea, but I am not sure it will work out of the box since the target is not an elasticsearch instance and I believe sense uses the kibana configuration to get the address of the server?

acchen97 replied:
Not sure how this would work technically, perhaps we can leave out the Sense portions for now.

suyograo replied:
+1 on leaving sense out. IMO most users of monitoring APIs are for operational health and not for developing. So they might just use `curl` or tie this into an existing monitoring solution

dedemorton replied:
@ph Yes, that's correct. The topic describes what users need to do to get it to work. I'm curious to hear what other people think. I'll gladly remove the Sense links if people think they are too confusing. They are a nice (easy-to-implement) extra, but not required. @suyograo Do you have an opinion?

suyograo replied:
@dedemorton +1 on leaving sense out. IMO most uses of monitoring APIs are for operational health and not for developing or learning (which is where Sense is invaluable). So operators might just use `curl` or tie this into an existing monitoring solution.

ph commented on docs/static/monitoring-apis.asciidoc:
```diff
@@ -0,0 +1,312 @@
+to the query string. This makes sense when the stats results are
+being consumed by a monitoring tool, rather than intended for human
+consumption.  The default for the `human` flag is
+`false`.
```
This is not exactly true, the API will always return the two formats the machine format and the human format in the JSON response. So you will get both `exists_time` and `exists_time_in_millis` in your response.

The human params currently only works the `hot_threads` api call and it currently do the following it will return a plain text string instead of a json document which make the response easier to read than a formatted json.

dedemorton replied:
@ph Is this true for Alpha1 or in general?

I was under the impression from my conversation with @suyograo and Alvin that these parameters would provide the same behavior that they provide in the Elasticsearch APIs (documented [here](https://www.elastic.co/guide/en/elasticsearch/reference/current/common-options.html)). 

Are you planning to add wider support for `?human` post-Alpha1? If so, I'm wondering if it would make sense to document the behavior as we'll deliver it in 5.0.0 and add an Alpha1 note indicating that the parameter is only supported for the hot threads API.

ph replied:
> Are you planning to add wider support for ?human post-Alpha1? If so, I'm wondering if it would make sense to document the behavior as we'll deliver it in 5.0.0 and add an Alpha1 note indicating that the parameter is only supported for the hot threads API.

I think this would be a great idea, since we will probably add more support to the `human` option.

ph commented on docs/static/monitoring-apis.asciidoc:
```diff
@@ -0,0 +1,312 @@
+
+experimental[]
+
+//QUESTION: Which of the params shown for the Elasticsearch hot_threads API also available here? (I'm talking about the threads, interval, type, and ignore_idle_threads parameters shown here: https://www.elastic.co/guide/en/elasticsearch/reference/2.2/cluster-nodes-hot-threads.html) I ask even though these parameters are not mentioned in issue #4446 because I tested ?threads, and it works, but wh ...
```
The current query string params support by this call are  the following.

`thread` => integer, default 3, limit the number of hot threads returned by the api, they ar ethe threads that consume the most resources.
`human` => boolean default false, return the result in plain text instead of the json format.
`ignore_idle_threads` => default true, will not return idle threads.

ph commented on docs/static/monitoring-apis.asciidoc:
```diff
@@ -0,0 +1,312 @@
+  } ]
+--------------------------------------------------
+
+//QUESTION: Do you think we should describe the fields that are returned by each API? I haven't done this because I feel that the fields are intuitively named and that the descriptions would be redundant. WDYT?
```
I agree I believe they are self explicit and we might add more fields so it would make the doc harder to keep in sync with the actual code.

suyograo replied:
+1, no need to describe individual fields

ph commented on docs/static/monitoring-apis.asciidoc:
```diff
@@ -0,0 +1,312 @@
+
+experimental[]
+
+//QUESTION: Is "root resource API" a good name for this API? 
```
I think this is an OK name, I wish naming thing was easier from time to time ;)
@purbon do you have a better name for this one?

suyograo replied:
Yeah, root resource is fine.

ph commented on docs/static/monitoring-apis.asciidoc:
```diff
@@ -0,0 +1,312 @@
+The root resource API retrieves general information about the Logstash instance, including
+the host name, version, and build hash.
+
+//QUESTION: Actually I tested the REST call on the latest build, the response did not include the build hash. Is this field still part of the output?
```
I think in the current version we don't return the `build_hash`, I have created an issue to discuss if we still want that feature or not.

See https://github.com/elastic/logstash/issues/4786

ph commented on docs/static/monitoring-apis.asciidoc:
```diff
@@ -0,0 +1,312 @@
+To retrieve all stats per pipeline, use the `_pipelines/stats` endpoint:
+
+//QUESTION: Should we mention that for Alpha 1 that the API will only show one pipeline, main? Actually, when I try the REST call, it returns an empty document. Is the _pipelines/stats endpoint even supported in Alpha 1? I'm using a very basic Logstash config. Do I need to configure something to see pipeline stats? 
+
```
The stats are dynamically created at runtime, if the pipeline don't do anything no stats will be collected. If you are using the `generator` input to send to elasticsearch you should get some metrics. Can you share you configuration?

ph replied:
I think the current expectation for people is that logstash only run one pipeline, so I think we don't need to explain that the stats will be from the main pipeline? This could change in the future and when it happen we can update the documentation to reflect the new reality.

ph commented on docs/static/monitoring-apis.asciidoc:
```diff
@@ -0,0 +1,312 @@
+
+Where `<types>` is optional and specifies the types of stats you want to return.
+
+//QUESTION: I'm assuming that I should be able to specify more than one type by using a comma separator, but when I specify more than one type, the call returns an empty document. Am I mistaken? For example, the following command returns {}: curl localhost:9600/_node/stats/events,jvm. I know that right now it doesn't make sense to specify the two separated by a comma because you get that by defau ...
```
This might be a bug @purbon can you double check that?

ph commented:
I think an initial review, I think I have answered all your questions. Good stuff! Thanks @dedemorton

acchen97 commented on docs/static/monitoring-apis.asciidoc:
```diff
@@ -0,0 +1,312 @@
+Logstash provides the following monitoring APIs to retrieve runtime metrics
+about Logstash:
+
+* <<hot-threads-api>>
```
Can we shift hot threads to the last item (under stats info)?

dedemorton replied:
@acchen97 Right now the APIs are listed in alphabetical order. One of my pet peeves about our documentation is that we don't always have a clear organizational pattern, which leaves it up to users to figure out where we've buried stuff. For reference info, I prefer to alphabetize the topics to make it easier for people to find stuff by name (which admittedly isn't important with a small set of topics). If we are planning to add a bunch of other monitoring APIs later, then I would prefer to keep the current organization and be consistent in all the APIs that we document. I'm open to your suggestion, though, if you feel strongly.

dedemorton replied:
@acchen97 Actually, ignore my previous comment. I'll move the section to the end for now and we can decide on the final organization when/if we have more APIs to worry about and the order matters more. :-)

acchen97 commented on docs/static/monitoring-apis.asciidoc:
```diff
@@ -0,0 +1,312 @@
+
+[source,js]
+--------------------------------------------------
+GET /_pipelines/stats/<types>
```
Let's remove this _pipelines option as its a future item.

dedemorton commented on docs/static/monitoring-apis.asciidoc:
```diff
@@ -0,0 +1,312 @@
+
+[source,js]
+--------------------------------------------------
+GET /_stats/jvm
```
Good catch. Fixed.

dedemorton commented:
@suyograo, @ph  @acchen97 : I've pushed changes based on your feedback. Let me know if the changes look OK and whether this is ready to merge into the main docs. 

I've refreshed the built files so you can see the doc in the right context: https://logstashreference.firebaseapp.com/monitoring.html (you might need to reload the page to see the changes)

suyograo commented:
@dedemorton looking good. What about a top level section "Common Options" we can then document `?pretty` and `?human` settings?

dedemorton commented:
@suyograo I've moved the section about common options to the main container topic (I think that's what you were asking me to do). Also, I've added links to each reference topic to point back to the Common Options section. I've refreshed the draft at https://logstashreference.firebaseapp.com/monitoring.html.

suyograo commented:
LGTM. Please merge

acchen97 commented:
Good work @dedemorton thanks! :+1:
